plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java

```diff
@@ -0,0 +1,570 @@
+
+            currentFile = null;
+            sliceInput = null;
+            bufferRetainedSize = 0;
+            inProgressReadFuture.cancel(true);
+            inProgressReadFuture = immediateVoidFuture(); // such that we don't retain reference to the buffer
+        }
+
+        private void fillBuffer()
+        {
+            if (currentFile == null || fileOffset == currentFile.getFileSize()) {
+                currentFile = sourceFiles.poll();
+                if (currentFile == null) {
+                    close();
+                    return;
+                }
+                fileOffset = 0;
+            }
+
+            byte[] buffer = new byte[bufferSize];
+            int bufferFill = 0;
+            if (sliceInput != null) {
+                int length = sliceInput.available();
+                sliceInput.readBytes(buffer, 0, length);
+                bufferFill += length;
+            }
+
+            ImmutableList.Builder<ListenableFuture<Void>> downloadFutures = ImmutableList.builder();
+            while (true) {
+                long fileSize = currentFile.getFileSize();
+                // Make sure Azure Blob Storage read request byte ranges align with block sizes for best performance
+                int readableBlocks = (buffer.length - bufferFill) / blockSize;
+                if (readableBlocks == 0) {
+                    if (buffer.length - bufferFill >= fileSize - fileOffset) {
+                        readableBlocks = 1;
+                    }
+                    else {
+                        break;
+                    }
+                }
+
+                BlockBlobAsyncClient blockBlobAsyncClient = blobServiceAsyncClient
+                        .getBlobContainerAsyncClient(getContainerName(currentFile.getFileUri()))
+                        .getBlobAsyncClient(getPath(currentFile.getFileUri()))
+                        .getBlockBlobAsyncClient();
+                Optional<SecretKey> secretKey = currentFile.getSecretKey();
+                if (secretKey.isPresent()) {
+                    blockBlobAsyncClient = blockBlobAsyncClient.getCustomerProvidedKeyAsyncClient(new CustomerProvidedKey(secretKey.get().getEncoded()));
+                }
+                for (int i = 0; i < readableBlocks && fileOffset < fileSize; ++i) {
+                    int length = (int) min(blockSize, fileSize - fileOffset);
+
+                    int finalBufferFill = bufferFill;
+                    FluentFuture<Void> downloadFuture = FluentFuture.from(toListenableFuture(blockBlobAsyncClient.downloadWithResponse(new BlobRange(fileOffset, (long) length), null, null, false).toFuture()))
+                            .transformAsync(response -> toListenableFuture(response.getValue().collectList().toFuture()), directExecutor())
+                            .transform(byteBuffers -> {
+                                int offset = finalBufferFill;
+                                for (ByteBuffer byteBuffer : byteBuffers) {
```

Comment:
Is it possible to have the client directly write bytes into a destination buffer (to avoid double copying?)

Reply:
I don't see an easy way, as Azure SDK doesn't provide any kind of callback/transformer interface to do so.